otherlibs/systhreads/st_pthreads.h

```diff
@@ -0,0 +1,320 @@
+       yield() or enter_blocking_section() call (or we see a spurious condvar
+       wakeup, which are rare at best.) */
+       pthread_cond_wait(&m->is_free, &m->lock);
+  } while (m->busy);
+
+  m->busy = 1;
+  atomic_fetch_add(&m->waiters, -1);
+
+  caml_acquire_domain_lock();
+
+  pthread_mutex_unlock(&m->lock);
+
+  return;
+}
+
+/* Triggered events */
+
+typedef struct st_event_struct {
+  pthread_mutex_t lock;         /* to protect contents */
+  int status;                   /* 0 = not triggered, 1 = triggered */
+  pthread_cond_t triggered;     /* signaled when triggered */
+} * st_event;
+
+
+static int st_event_create(st_event * res)
+{
+  int rc;
+  st_event e = caml_stat_alloc_noexc(sizeof(struct st_event_struct));
+  if (e == NULL) return ENOMEM;
+  rc = pthread_mutex_init(&e->lock, NULL);
+  if (rc != 0) { caml_stat_free(e); return rc; }
+  rc = pthread_cond_init(&e->triggered, NULL);
+  if (rc != 0)
+  { pthread_mutex_destroy(&e->lock); caml_stat_free(e); return rc; }
+  e->status = 0;
+  *res = e;
+  return 0;
+}
+
+static int st_event_destroy(st_event e)
+{
+  int rc1, rc2;
+  rc1 = pthread_mutex_destroy(&e->lock);
+  rc2 = pthread_cond_destroy(&e->triggered);
+  caml_stat_free(e);
+  return rc1 != 0 ? rc1 : rc2;
+}
+
+static int st_event_trigger(st_event e)
+{
+  int rc;
+  rc = pthread_mutex_lock(&e->lock);
+  if (rc != 0) return rc;
+  e->status = 1;
+  rc = pthread_mutex_unlock(&e->lock);
+  if (rc != 0) return rc;
+  rc = pthread_cond_broadcast(&e->triggered);
+  return rc;
+}
+
+static int st_event_wait(st_event e)
+{
+  int rc;
+  rc = pthread_mutex_lock(&e->lock);
+  if (rc != 0) return rc;
+  while(e->status == 0) {
+    rc = pthread_cond_wait(&e->triggered, &e->lock);
+    if (rc != 0) return rc;
+  }
+  rc = pthread_mutex_unlock(&e->lock);
+  return rc;
+}
+
+/* The tick thread: interrupt the domain periodically to force preemption  */
+
+static void * caml_thread_tick(void * arg)
+{
+  caml_domain_state *domain;
+  uintnat *domain_id = (uintnat *) arg;
+
+  caml_init_domain_self(*domain_id);
+  domain = Caml_state;
+
+  caml_domain_set_name("Tick");
+  while(! atomic_load_acq(&Tick_thread_stop)) {
+    st_msleep(Thread_timeout);
+
+    atomic_store_rel((atomic_uintnat*)&domain->requested_external_interrupt, 1);
+    caml_interrupt_self();
+  }
+  return NULL;
+}
+
+/* "At fork" processing */
+
+#if defined(__ANDROID__)
+/* Android's libc does not include declaration of pthread_atfork;
+   however, it implements it since API level 10 (Gingerbread).
+   The reason for the omission is that Android (GUI) applications
+   are not supposed to fork at all, however this workaround is still
+   included in case OCaml is used for an Android CLI utility. */
+int pthread_atfork(void (*prepare)(void), void (*parent)(void),
+                   void (*child)(void));
+#endif
+
+static int st_atfork(void (*fn)(void))
+{
+  caml_atfork_hook = fn;
+  return 0;
+}
```

Comment:
The Android declaration is useless now that we don't call `pthread_atfork` and should be removed.

 `st_atfork` could also be removed, just by inlining `caml_atfork_hook = caml_thread_reinitialize;` in st_stubs.c, function `caml_thread_initialize`.

Reply:
Done!